File: src/providers/cache.py
"""Filesystem-based cache for last-good usage data."""

import json
import time
from pathlib import Path
from typing import Dict, Optional

from .base import DataSource, FetchResult, UsageProvider


class UsageCache(UsageProvider):
    """Caches last successful usage data to filesystem."""

    CACHE_DIR = Path.home() / ".cache" / "usage-cli"
    STALE_THRESHOLD = 300  # 5 minutes

    def __init__(self, service: str):
        """
        Initialize cache for a service.

        Args:
            service: Service name ('claude' or 'codex')
        """
        super().__init__()
        self.service = service
        self.source_type = DataSource.CACHE
        self.cache_file = self.CACHE_DIR / f"{service}.json"

        # Ensure cache directory exists
        self.CACHE_DIR.mkdir(parents=True, exist_ok=True)

    def is_available(self) -> bool:
        """Check if cache file exists."""
        return self.cache_file.exists()

    def fetch(self) -> FetchResult:
        """Fetch data from cache."""
        timestamp = time.time()

        if not self.cache_file.exists():
            return FetchResult(
                metrics=None,
                source=self.source_type,
                timestamp=timestamp,
                error="Cache file does not exist"
            )

        try:
            with open(self.cache_file, 'r') as f:
                data = json.load(f)

            cached_timestamp = data.get("timestamp", 0)
            metrics = data.get("metrics")

            # Check if cache is stale
            age = timestamp - cached_timestamp
            is_stale = age > self.STALE_THRESHOLD

            return FetchResult(
                metrics=metrics,
                source=self.source_type,
                timestamp=cached_timestamp,
                error=None,
                stale=is_stale
            )

        except (json.JSONDecodeError, OSError) as e:
            return FetchResult(
                metrics=None,
                source=self.source_type,
                timestamp=timestamp,
                error=f"Failed to read cache: {str(e)}"
            )

    def store(self, metrics: Dict, timestamp: float):
        """
        Store metrics to cache.

        Args:
            metrics: Metrics dict to cache
            timestamp: Timestamp of the data
        """
        try:
            data = {
                "timestamp": timestamp,
                "metrics": metrics,
            }

            with open(self.cache_file, 'w') as f:
                json.dump(data, f, indent=2)

        except OSError:
            # Silently fail, cache is best-effort
            pass

    def clear(self):
        """Clear the cache file."""
        try:
            if self.cache_file.exists():
                self.cache_file.unlink()
        except OSError:
            pass

File: packages/core/src/providers/cache.ts
/**
 * Filesystem-based cache for last-good usage data.
 * Port of src/providers/cache.py
 */

import { homedir } from "os";
import { join } from "path";
import { existsSync, mkdirSync, readFileSync, writeFileSync, unlinkSync } from "fs";
import { DataSource } from "../types.js";
import type { FetchResult, UsageProvider, MetricsDict } from "../types.js";

export class UsageCache implements UsageProvider {
  static readonly CACHE_DIR = join(homedir(), ".cache", "usage-cli");
  static readonly STALE_THRESHOLD = 300; // 5 minutes

  name = "UsageCache";
  sourceType = DataSource.CACHE;

  private cacheFile: string;

  constructor(service: string) {
    this.cacheFile = join(UsageCache.CACHE_DIR, `${service}.json`);

    // Ensure cache directory exists
    if (!existsSync(UsageCache.CACHE_DIR)) {
      mkdirSync(UsageCache.CACHE_DIR, { recursive: true });
    }
  }

  isAvailable(): boolean {
    return existsSync(this.cacheFile);
  }

  async fetch(): Promise<FetchResult> {
    const timestamp = Date.now() / 1000;

    if (!existsSync(this.cacheFile)) {
      return {
        metrics: null,
        source: this.sourceType,
        timestamp,
        error: "Cache file does not exist",
        stale: false,
      };
    }

    try {
      const raw = readFileSync(this.cacheFile, "utf-8");
      const data = JSON.parse(raw) as { timestamp: number; metrics: MetricsDict };

      const cachedTimestamp = data.timestamp ?? 0;
      const metrics = data.metrics;

      const age = timestamp - cachedTimestamp;
      const isStale = age > UsageCache.STALE_THRESHOLD;

      return {
        metrics,
        source: this.sourceType,
        timestamp: cachedTimestamp,
        error: null,
        stale: isStale,
      };
    } catch (e) {
      return {
        metrics: null,
        source: this.sourceType,
        timestamp,
        error: `Failed to read cache: ${e instanceof Error ? e.message : String(e)}`,
        stale: false,
      };
    }
  }

  store(metrics: MetricsDict, timestamp: number): void {
    try {
      const data = { timestamp, metrics };
      writeFileSync(this.cacheFile, JSON.stringify(data, null, 2));
    } catch {
      // Silently fail, cache is best-effort
    }
  }

  clear(): void {
    try {
      if (existsSync(this.cacheFile)) {
        unlinkSync(this.cacheFile);
      }
    } catch {
      // Silently fail
    }
  }
}
